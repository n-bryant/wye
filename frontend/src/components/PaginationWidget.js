import React from "react";
import PropTypes from "prop-types";

import { withStyles } from "@material-ui/core/styles";
import createClassNameHelper from "@n_bryant/classnames-helper";
import JSS_CLASS_NAME_PREFIX from "../../lib/classNamePrefix";

import Icon from "@mdi/react";
import {
  mdiChevronLeft,
  mdiChevronRight,
  mdiPageFirst,
  mdiPageLast
} from "@mdi/js";

import Button from "@material-ui/core/Button";
import Typography from "@material-ui/core/Typography";
import styles from "./PaginationWidget.styles";

/**
 * renders a pagination widget with prev/next buttons and an first/last page buttons
 */
export const PaginationWidget = props => {
  const classnames = PaginationWidget.classnames(props);
  const { currentPage, setCurrentPage, totalPages } = props;

  // paddle disabled logic
  const prevPaddleDisabled = currentPage === 1;
  const nextPaddleDisabled = currentPage === totalPages;

  return (
    <div className={classnames.root()}>
      {totalPages > 2 && (
        <Button
          className={classnames.element("paddle")}
          disabled={prevPaddleDisabled}
          onClick={() => setCurrentPage(1)}
        >
          <Icon
            className={classnames.element("paddleIcon", {
              disabled: prevPaddleDisabled
            })}
            path={mdiPageFirst}
          />
        </Button>
      )}
      <Button
        className={classnames.element("paddle")}
        disabled={prevPaddleDisabled}
        onClick={() => setCurrentPage(currentPage - 1)}
      >
        <Icon
          className={classnames.element("paddleIcon", {
            disabled: prevPaddleDisabled
          })}
          path={mdiChevronLeft}
        />
      </Button>
      <Typography variant="body2">{`${currentPage} of ${totalPages}`}</Typography>
      <Button
        className={classnames.element("paddle")}
        disabled={nextPaddleDisabled}
        onClick={() => setCurrentPage(currentPage + 1)}
      >
        <Icon
          className={classnames.element("paddleIcon", {
            disabled: nextPaddleDisabled
          })}
          path={mdiChevronRight}
        />
      </Button>
      {totalPages > 2 && (
        <Button
          className={classnames.element("paddle")}
          disabled={nextPaddleDisabled}
          onClick={() => setCurrentPage(totalPages)}
        >
          <Icon
            className={classnames.element("paddleIcon", {
              disabled: nextPaddleDisabled
            })}
            path={mdiPageLast}
          />
        </Button>
      )}
    </div>
  );
};
PaginationWidget.classnames = createClassNameHelper(
  `${JSS_CLASS_NAME_PREFIX}PaginationWidget`
);
PaginationWidget.propTypes = {
  // styles to apply
  classes: PropTypes.shape({
    root: PropTypes.string,
    paddlesWrapper: PropTypes.string,
    paddle: PropTypes.string,
    paddleIcon: PropTypes.string,
    paddleIconDisabled: PropTypes.string,
    pageIndicator: PropTypes.string,
    pagingValuesContainer: PropTypes.string,
    goToPageInput: PropTypes.string
  }),
  // the current page number
  currentPage: PropTypes.number,
  // handler to set the current page
  setCurrentPage: PropTypes.func,
  // the count of total pages
  totalPages: PropTypes.number
};
PaginationWidget.defaultProps = {
  classes: {},
  currentPage: 1,
  setCurrentPage: () => {}
};

export default withStyles(styles)(PaginationWidget);
