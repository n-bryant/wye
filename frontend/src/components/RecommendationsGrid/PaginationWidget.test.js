import { PaginationWidget } from "./PaginationWidget";
import {
  mdiChevronLeft,
  mdiChevronRight,
  mdiPageFirst,
  mdiPageLast
} from "@mdi/js";

describe("PaginationWidget", () => {
  const props = {
    classes: {
      root: "root",
      paddlesWrapper: "paddlesWrapper",
      paddle: "paddle",
      paddleIcon: "paddleIcon",
      paddleIconDisabled: "paddleIconDisabled",
      pageIndicator: "pageIndicator",
      pagingValuesContainer: "pagingValuesContainer",
      goToPageInput: "goToPageInput"
    },
    currentPage: 1,
    setCurrentPage: jest.fn(),
    totalPages: 3
  };

  let wrapper;
  beforeEach(() => {
    wrapper = shallow(<PaginationWidget {...props} />);
  });

  it("should render successfully", () => {
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it("should set the first page button and prev page button to disabled if the current page is the first page", () => {
    expect(
      wrapper
        .findWhere(n => n.hasClass(props.classes.paddle))
        .at(0)
        .prop("disabled")
    ).toBeTruthy();
    expect(
      wrapper
        .findWhere(n => n.hasClass(props.classes.paddle))
        .at(1)
        .prop("disabled")
    ).toBeTruthy();
  });

  it("should set disabled styling to the paddle icon of the prev and first page buttons when the button is disabled", () => {
    expect(
      wrapper
        .findWhere(n => n.prop("path") === mdiPageFirst)
        .hasClass(props.classes.paddleIconDisabled)
    ).toBeTruthy();
    expect(
      wrapper
        .findWhere(n => n.prop("path") === mdiChevronLeft)
        .hasClass(props.classes.paddleIconDisabled)
    ).toBeTruthy();
  });

  it("should set the last page button and next page button to disabled if the current page is the last page", () => {
    wrapper.setProps({
      currentPage: props.totalPages
    });
    wrapper.update();
    expect(
      wrapper
        .findWhere(n => n.hasClass(props.classes.paddle))
        .at(2)
        .prop("disabled")
    ).toBeTruthy();
    expect(
      wrapper
        .findWhere(n => n.hasClass(props.classes.paddle))
        .at(3)
        .prop("disabled")
    ).toBeTruthy();
  });

  it("should set disabled styling to the paddle icon of the next and last page buttons when the button is disabled", () => {
    wrapper.setProps({
      currentPage: props.totalPages
    });
    wrapper.update();
    expect(
      wrapper
        .findWhere(n => n.prop("path") === mdiPageLast)
        .hasClass(props.classes.paddleIconDisabled)
    ).toBeTruthy();
    expect(
      wrapper
        .findWhere(n => n.prop("path") === mdiChevronRight)
        .hasClass(props.classes.paddleIconDisabled)
    ).toBeTruthy();
  });
});
