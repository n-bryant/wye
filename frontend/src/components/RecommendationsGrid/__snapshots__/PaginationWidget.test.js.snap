// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PaginationWidget should render successfully 1`] = `
<div
  className="wye-PaginationWidget root"
>
  <WithStyles(ForwardRef(Button))
    className="wye-PaginationWidget__paddle paddle"
    disabled={true}
    onClick={[Function]}
  >
    <Icon
      className="wye-PaginationWidget__paddleIcon wye-PaginationWidget__paddleIcon--disabled paddleIcon paddleIconDisabled"
      color={null}
      horizontal={false}
      path="M18.41,16.59L13.82,12L18.41,7.41L17,6L11,12L17,18L18.41,16.59M6,6H8V18H6V6Z"
      rotate={0}
      size={null}
      spin={false}
      vertical={false}
    />
  </WithStyles(ForwardRef(Button))>
  <WithStyles(ForwardRef(Button))
    className="wye-PaginationWidget__paddle paddle"
    disabled={true}
    onClick={[Function]}
  >
    <Icon
      className="wye-PaginationWidget__paddleIcon wye-PaginationWidget__paddleIcon--disabled paddleIcon paddleIconDisabled"
      color={null}
      horizontal={false}
      path="M15.41,16.58L10.83,12L15.41,7.41L14,6L8,12L14,18L15.41,16.58Z"
      rotate={0}
      size={null}
      spin={false}
      vertical={false}
    />
  </WithStyles(ForwardRef(Button))>
  <WithStyles(ForwardRef(Typography))
    variant="body2"
  >
    Page 1 of 3
  </WithStyles(ForwardRef(Typography))>
  <WithStyles(ForwardRef(Button))
    className="wye-PaginationWidget__paddle paddle"
    disabled={false}
    onClick={[Function]}
  >
    <Icon
      className="wye-PaginationWidget__paddleIcon paddleIcon"
      color={null}
      horizontal={false}
      path="M8.59,16.58L13.17,12L8.59,7.41L10,6L16,12L10,18L8.59,16.58Z"
      rotate={0}
      size={null}
      spin={false}
      vertical={false}
    />
  </WithStyles(ForwardRef(Button))>
  <WithStyles(ForwardRef(Button))
    className="wye-PaginationWidget__paddle paddle"
    disabled={false}
    onClick={[Function]}
  >
    <Icon
      className="wye-PaginationWidget__paddleIcon paddleIcon"
      color={null}
      horizontal={false}
      path="M5.59,7.41L10.18,12L5.59,16.59L7,18L13,12L7,6L5.59,7.41M16,6H18V18H16V6Z"
      rotate={0}
      size={null}
      spin={false}
      vertical={false}
    />
  </WithStyles(ForwardRef(Button))>
</div>
`;
